lib/puppet/application/resource.rb

```diff
@@ -225,6 +225,7 @@ def create_default_environment
   def find_or_save_resources(type, name, params)
     key = local_key(type, name)
 
+    Puppet.push_context({:stringify_rich => true})
```

Comment:
We try to only use `push_context` when bootstrapping puppet, and then use `Puppet.override(stringify_rich: true) do ... end` so that the context remains balanced when the method returns.